### PR TITLE
feat(behavior_path_planner): better avoidance drivable areas extension in behavior path planning

### PR DESCRIPTION
**Pull request by zulfaqar-azmi-t4**
## Related Issue(required)

#234 

## Description(required)

This PR add adjacent lane check for left side and right side of input lanelet.
by implementing the check function, it is possible to check whether there exist any lane at the left side or right side of ego vehicle lane.

This PR is a port from .iv. It is divided into two parts,

1. route_handler(no change to behavior)
2. behavior_path_planner (shift_length computation feature)

The feature implement adjacent lanelet search to prevent planner from generating avoidance path that extend towards non-lane areas like below

image

It also improve drivable area extension by appending lanelet that is nearby obstacle.
## Review Procedure(required)

Use the scenario files included in the following [tier4/AutonomousDrivingScenarios#287](https://github.com/tier4/AutonomousDrivingScenarios/pull/287). 
[new avoidance.param.yaml](https://github.com/tier4/autoware_launcher/pull/496) is needed.

## Related PR(optional)

#285 
https://github.com/tier4/autoware_launch/pull/171

this feature is tested in autoware.iv this scenario simulator
.iv
Azu : [tier4/autoware.iv#2383](https://github.com/tier4/autoware.iv/pull/2383)
Murooka : [tier4/autoware.iv#1839](https://github.com/tier4/autoware.iv/pull/1839)
[tier4/AutonomousDrivingScenarios#287](https://github.com/tier4/AutonomousDrivingScenarios/pull/287)
[new avoidance.param.yaml](https://github.com/tier4/autoware_launcher/pull/496)
## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [ ] Read [commit-guidelines][commit-guidelines]
- [ ] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by taikitanaka3** [comment]
@zulfaqar-azmi-t4 
can you remove route handler files and rebase to latest commit ? https://github.com/autowarefoundation/autoware.universe/pull/285

**Comment by zulfaqar-azmi-t4** [comment]
> can you remove route handler files and rebase to latest commit ? #285
Yes, I'm on it.

**Comment by zulfaqar-azmi-t4** [comment]
@taikitanaka3 
I fixed the spell-check in the `.iv`, but I didn't include them in this PR since it doesn't related to the feature. So, do I need to fix them here?

**Comment by taikitanaka3** [comment]
@zulfaqar-azmi-t4 
you can make another pr to fix spell check but please don't include spell check to this PR.

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp [hunk 1]
```diff
@@ -242,6 +265,7 @@ struct DebugData
 {
   std::shared_ptr<lanelet::ConstLanelets> expanded_lanelets;
   std::shared_ptr<lanelet::ConstLanelets> current_lanelets;
+  std::shared_ptr<lanelet::ConstLineStrings3d> linestring_wrt_overhang;
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp:268]: @zulfaqar-azmi-t4 
can you make this name more clear? 
for example linestring_to_farthest_overhang

Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp:268]: is `farthest_linestring_from_overhang` is okay?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -221,25 +227,74 @@ ObjectDataArray AvoidanceModule::calcAvoidanceTargetObjects(
...
+        object_data.to_road_shoulder_distance =
+          distance2d(to2D(overhang_basic_pose), to2D(target_right_line.basicLineString()));
+        debug_linestring.push_back(target_right_line);
+      }
+    }
 
     DEBUG_PRINT(
-      "set object_data: longitudinal = %f, lateral = %f, largest_overhang = %f",
-      object_data.longitudinal, object_data.lateral, object_data.overhang_dist);
+      "set object_data: longitudinal = %f, lateral = %f, largest_overhang = %f, "
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:280]:
```suggestion
      "set object_data: longitudinal = %f, lateral = %f, largest_overhang = %f", 
```

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp (hunk 1 above)
Comment (taikitanaka3) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp:268]: good

Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/avoidance/avoidance_module_data.hpp:268]: I'm on it.

**Comment by taikitanaka3** [comment]
-  Case avoidance within lanelet
image
-  Don't avoid if ego vehile can't avoid interrupting object within lanelet 
image

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (taikitanaka3) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: @zulfaqar-azmi-t4 
if you press commit suggestion you can apply it alternetively

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md [hunk 3]
```diff
@@ -238,6 +238,148 @@ The shift points are modified by a filtering process in order to get the expecte
...
+  - It is recommended to set the value to more than half of the ego vehicle's width.
+- The `road_shoulder_safety_margin` will prevent the module from generating a path that might cause the vehicle to go too near the road shoulder.
+
+![fig1](./image/shift_length_parameters.png)
+
+The shift length is subjected to the following constraints.
+
+<!-- spell-checker:disable -->
+
+![shift_length_equation](./image/shift_length_equation.png)
```
Comment (takayuki5168) [inline planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:270]: You can write mathematical formulation with math jax. See an example in readme of obstacle avoidance planner.

**Comment by zulfaqar-azmi-t4** [comment]
@taikitanaka3 
Just to confirm, if I'm not mistaken the [in the second image](https://github.com/autowarefoundation/autoware.universe/pull/287#issuecomment-1019781761), the vehicle will stop right before the obstacle. Please let me know if it didn't happen in your case.

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md [hunk 4]
```diff
@@ -238,6 +238,148 @@ The shift points are modified by a filtering process in order to get the expecte
...
+
+<!--<img src="https://latex.codecogs.com/gif.latex?\textrm{shift&space;length}=\left\{\begin{matrix}&space;d_{lcsb}&plus;d_{lcm}&plus;\frac{1}{2}(W_{ego})&\textrm{if}&\left(d_{lcsb}&plus;d_{lcm}&plus;W_{ego}&plus;d_{rssm}\right)<d_{trsd}\\&space;0&space;&&space;\textrm{if}&\left(d_{lcsb}&plus;d_{lcm}&plus;W_{ego}&plus;d_{rssm}\right)\geq&space;d_{trsd}&space;\end{matrix}\right." title="shift_length_computation" /> -->
```
Comment (takayuki5168) [inline planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:272]: Should remove shift_lefngth_computation.png? Even if you wanna use the figure in readme, you'd better use plantuml.
See an example in readme of behavior velocity, obstacle stop, ... (used in most planning packages' readme)

Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:272]: Might be the way I include the equation is wrong. I can try remove the `png` and uncomment the equation.

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md (hunk 3 above)
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:270]: I'll try to change the latex convention to math jax.

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md (hunk 4 above)
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/behavior_path_planner_avoidance-design.md]: Done. I've update the equations.

**Comment by taikitanaka3** [comment]
> @taikitanaka3 Just to confirm, if I'm not mistaken the [in the second image](https://github.com/autowarefoundation/autoware.universe/pull/287#issuecomment-1019781761), the vehicle will stop right before the obstacle. Please let me know if it didn't happen in your case.

I set motion planning markers off to see clearly for this PR review so vehicle will be stop in front of obstacle.

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md (hunk 3 above)
Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/behavior_path_planner_avoidance-design.md]: Done. I've update the equations.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 5]
```diff
@@ -1584,22 +1662,195 @@ double AvoidanceModule::getLeftShiftBound() const
...
-  right_extend -= (right_extend < -THRESHOLD) ? margin : 0.0;
-  left_extend += (left_extend > THRESHOLD) ? margin : 0.0;
+        return false;
+      }();
+      if (update_extended_lanelets) {
+        continue;
+      }
+    }
 
-  const auto extended_lanelets = lanelet::utils::getExpandedLanelets(
```
Comment (taikitanaka3) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1601]: @zulfaqar-azmi-t4 
Note:
removed extend lanelets by constant margin

Comment (zulfaqar-azmi-t4) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:1601]: Yes, this removes [fix lanelet expansion](https://github.com/autowarefoundation/autoware.universe/pull/287/files#:~:text=constexpr%20double%20extend_margin%20%3D%200.5%3B) in favor of directly insert the lanelet to 
extended drivable areas list.

**Review by taikitanaka3 — APPROVED** [review]
LGTM
image

**Review by taikitanaka3 — APPROVED** [review]

**Comment by taikitanaka3** [comment]
@zulfaqar-azmi-t4 
can you merge autoware_launch after this PR is merged?

**Comment by zulfaqar-azmi-t4** [comment]
> @zulfaqar-azmi-t4 can you merge autoware_launch after this PR is merged?

Yes, sure. I will merge them.

**Pull request merged**